### DISCUSSION TITLE
Mock a base class function

### DISCUSSION THREAD

**Discussion by cole701** [google/googletest#4572] [Community Help]
How do you mock a function that is part of a base class 3 levels of inheritance deep?

**Comment by AeroSW** [comment]
I assume you would mock it like you extend it.

```
class MockThirdlyDerivedStruct : public ThirdlyDerivedStruct {
   MOCK_METHOD(ret_val, func_name, (params...), (override);
};
```
As long as the method is `protected` scope, you should be able to access it from a derived class.  You can also make the method `virtual` in the `Base` or `...Derived` to let your compiler generate a Virtual Lookup Table entry for the method.

**Comment by cole701** [comment]
Thank you for the response. I thought the whole discussion group was dead!!

The base class is a Gtk library class and the function that I am attempting
to Mock is run().

Rich

**Reply by AeroSW** [reply]
So if you are using non-virtual polymorphism, I recommend using one of the Global Mocking Libraries, currently and personally I am trying [JOMOCK](https://github.com/jonah512/jomock) which has a non-virtual polymorphism example with GTest.  I am using JOMOCK for its static and global method mocking.

**Comment by cole701** [comment]
Kenneth,

Part of my problem is that my tool set is limited. I am on a closed
government system that is only allowed to have certain authorized
applications and libraries installed. So I have to stick with the
official applications that have been approved and that is gtest and gmock.
Additionally we are also currently stuck at version 1.8 for both until a
newer revision is authorized. So this is why I am trying to get this to
work with gmock.

Rich

**Reply by AeroSW** [reply]
I can't really help you there.  I know what that is like as well.  I used to work for SAIC.  GMock and GTest aren't designed to work with "legacy" global functions and static methods.  Something you can do though is look into the type of libraries (Gmock Global and JoMock) to see how they implemented their utilities, and "write" your own variation of their code as part of your code-base.

If you follow **GMock's** recommendations, they want you to write C++ as you would write Java, and use wrappers around those functions, so in your code, instead of directly calling `run` from your parent implementation, have a `virtual void runApp` method that calls the `run` method in your class, then stub/mock the calling of `runApp` rather than worrying about `run` itself.  You'll lose a little on your overall code coverage, but you would at least get away with not having to worry about mocking the original base's `run` method.

*Code below tested on [[OnlineGDB](https://www.onlinegdb.com/online_c++_compiler)]*
```
#include <iostream>

struct Base {
    void run() {
        std::cout << "Run called\n";
    }
};
struct Derived : public Base { // <-- This is the class you would create the mocked extension of.
    virtual void appRun() = 0; // <-- This is the method that would be mocked.
};
// The class below is what your software actually uses, and when testing this class specifically (not the section of code
// where this class is injected/used), you would just ignore `appRun` in your tests, since, it doesn't have any unique
// untested functional usage.  Assume the library where the Base class is from is already testing `run()`, so you shouldn't
// need to worry about skipping the one method that is wrapping its usage.
struct SecondlyDerived : public Derived {
    virtual void appRun() override {
        std::cout << "Calling Base run()\n";
        run();
    }
};

int main()
{
    std::cout<<"Hello World";
    SecondlyDerived sd;
    sd.appRun();
    return 0;
}
```
